src/serac/physics/utilities/functional/tests/functional_comparisons.cpp

```diff
@@ -113,18 +122,45 @@ void functional_test(mfem::ParMesh& mesh, H1<p> test, H1<p> trial, Dimension<dim
   // Compute the gradient using functional
   mfem::Operator& grad2 = residual.GetGradient(U);
 
+  // Test fully assembled matrix
+  mfem::Array<int> dofs;
+  fespace.GetElementDofs(0, dofs);
+  mfem::Vector K_e(mesh.GetNE() * dofs.Size() * fespace.GetVDim() * dofs.Size() * fespace.GetVDim());
+  K_e = 0.;
+  {
+    SERAC_PROFILE_SCOPE(concat("functional_gradientMatrix", postfix).c_str());
+    residual.GradientMatrix(K_e);
+  }
+
+  serac::mfem_ext::AssembledSparseMatrix A_serac_mat(fespace, fespace, mfem::ElementDofOrdering::LEXICOGRAPHIC);
+  {
+    SERAC_PROFILE_SCOPE(concat("AssembledSparseMatrix_FillData", postfix).c_str());
+    A_serac_mat.FillData(K_e);
+  }
+
+  A_serac_mat.Finalize();
+
+  std::unique_ptr<mfem::HypreParMatrix> J2(
+      SERAC_PROFILE_EXPR(concat("functional_gradParAssemble", postfix).c_str(), A.ParallelAssemble(&A_serac_mat)));
```

Comment:
will these intermediate quantities (K_e, A_serac_mat) be generally useful, or should `Functional::GradientMatrix` just return the HypreParMatrix itself?

Reply:
I think it depends on how many times you think you're going to re-assemble the matrix. Here I've assumed we're going to possibly assemble the same sparsity pattern /connectivity multiple times, so I would claim that `A_serac_mat` is more of a permanent quantity that is updated with `K_e`. I'm not sure, but I think HypreParMatrix needs to be possibly be "recomputed" each time the rank-local matrix is assembled. So I think that is the "temporary" quantity that we actually use to compute with.